File: HW8/8.2/8.2/tree.cpp
#include "tree.h"

struct Node
{
	char operation = '\0';
	int operand = 0;
	Node *leftChild = nullptr;
	Node *rightChild = nullptr;
};

struct Tree
{
	Node *root = nullptr;
};

Tree *createTree()
{
	return new Tree;
}

bool isOperation(char symbol)
{
	return symbol == '+' || symbol == '-' || symbol == '*' || symbol == '/';
}

Node *newNode(char *string, int &positionNumber)
{
	positionNumber++;
	while (string[positionNumber] == ' ' || string[positionNumber] == '(' || string[positionNumber] == ')')
	{
		positionNumber++;
	}

	Node *node = new Node;
	if (isOperation(string[positionNumber]))
	{
		node->operation = string[positionNumber];
		node->leftChild = newNode(string, positionNumber);
		node->rightChild = newNode(string, positionNumber);
	}
	else
	{
		node->operand = string[positionNumber] - '0';
	}
	return node;
}

Tree *build(char *string)
{
	int positionNumber = 0;
	Tree *tree = createTree();
	while (string[positionNumber] == ' ' || string[positionNumber] == '(')
	{
		positionNumber++;
	}
	tree->root = new Node;
	tree->root->operation = string[positionNumber];
	tree->root->leftChild = newNode(string, positionNumber);
	tree->root->rightChild = newNode(string, positionNumber);
	return tree;
}



void deleteChildren(Node *node)
{
	if (node == nullptr)
	{
		return;
	}
	deleteChildren(node->leftChild);
	deleteChildren(node->rightChild);
	delete node;
}

void deleteTree(Tree *tree)
{
	deleteChildren(tree->root);
	delete tree;
}

File: Semester2/Homeworks/HW4/Task1/Task1/Tree.cs
using System;
using Task1.Operations;

namespace Task1
{
    /// <summary>
    /// Abstract syntax tree.
    /// </summary>
    public class Tree
    {
        private INode root;

        /// <summary>
        /// Initializes a new instance of the <see cref="Tree"/> class.
        /// </summary>
        /// <param name="expression">Prefix expression</param>
        public Tree(string expression)
        {
            Build(expression);
        }

        private void Build(string expression)
        {
            var values = expression.Split(new char[] { ' ', '(', ')' }, StringSplitOptions.RemoveEmptyEntries);
            var position = -1;
            root = NewNode();

            INode NewNode()
            {
                position++;
                if (char.IsDigit(values[position], 0))
                {
                    return new Operand() { Value = int.Parse(values[position]) };
                }

                Operation newNode = values[position] switch
                {
                    "+" => new Addition(),
                    "-" => new Subtraction(),
                    "*" => new Multiplication(),
                    "/" => new Division(),
                    _ => null
                };

                newNode.LeftChild = NewNode();
                newNode.RightChild = NewNode();
                return newNode;
            }
        }

        /// <summary>
        /// Prints tree.
        /// </summary>
        public void Print() => root?.Print();

        /// <summary>
        /// Calculates the value of an expression by tree.
        /// </summary>
        /// <returns>Result</returns>
        public int Calculate() => root.Calculate();
    }
}
